### PR TITLE
Improve module variable taint propagation

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
Improve taint checking to global variable assignment to structured values.

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/19143

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java [hunk 1]
```diff
@@ -587,22 +587,22 @@ private void visitAssignment(BLangExpression varRefExpr, TaintedStatus varTainte
             if (varTaintedStatus == TaintedStatus.TAINTED && varRefExpr instanceof BLangVariableReference) {
                 BLangVariableReference varRef = (BLangVariableReference) varRefExpr;
                 if (isMutableVariable(varRef) && isGlobalVarOrServiceVar(varRef) && !isMarkedTainted(varRef)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java]: We seem to be repeating the `isStructuredAccessOnVariableReference()` check in `isMutableVariable()` and `isGlobalVarOrServiceVar()`. Can we reconsider the checks?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java [hunk 2]
```diff
@@ -587,22 +587,22 @@ private void visitAssignment(BLangExpression varRefExpr, TaintedStatus varTainte
...
+                    if (varRef.symbol != null && varRef.symbol.type.tag == TypeTags.OBJECT) {
+                        addTaintError(pos, getVariableName(varRef),
                                 DiagnosticCode.TAINTED_VALUE_PASSED_TO_MODULE_OBJECT);
                     } else {
-                        addTaintError(pos, varRef.symbol.name.value,
+                        addTaintError(pos, getVariableName(varRef),
                                 DiagnosticCode.TAINTED_VALUE_PASSED_TO_GLOBAL_VARIABLE);
                     }
                     return;
                 } else if (varRef.symbol != null && varRef.symbol.closure && varTaintedStatus == TaintedStatus.TAINTED
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java]: There's an IDE warning for `varTaintedStatus == TaintedStatus.TAINTED` --> always true.

Shall we fix this too?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java [hunk 3]
```diff
@@ -661,14 +661,30 @@ private boolean notInSameScope(BLangVariableReference varRefExpr, SymbolEnv env)
...
+            return isMarkedTainted((BLangVariableReference) ((BLangAccessExpression) varRef).expr);
+        }
         return ((BVarSymbol) varRef.symbol).taintabilityAllowance == BVarSymbol.TaintabilityAllowance.TAINTED;
     }
 
+    private boolean isStructuredAccessOnVariableReference(BLangVariableReference variableReference) {
+        return (variableReference.getKind() == NodeKind.INDEX_BASED_ACCESS_EXPR
+                || variableReference.getKind() == NodeKind.FIELD_BASED_ACCESS_EXPR
+                || variableReference.getKind() == NodeKind.XML_ATTRIBUTE_ACCESS_EXPR)
+                && ((BLangAccessExpression) variableReference).expr instanceof BLangVariableReference;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java]: Do we need the check for `((BLangAccessExpression) variableReference).expr instanceof BLangVariableReference`?

Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java]: I was under wrong impression about BLangVariableReference. Will remove it

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java (hunk 1 above)
Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java]: I'm not quite sure if I'm getting your idea, what we want it to check the `expr` part of AccessExpressions

File: tests/jballerina-integration-test/src/test/resources/grpc/src/grpcservices/03_bidirectional_streaming_service.bal [hunk 4]
```diff
@@ -33,7 +33,7 @@ listener grpc:Listener ep3 = new (9093, {
       }
   });
 
-map<grpc:Caller> connectionsMap = {};
+@tainted map<grpc:Caller> connectionsMap = {};
```
[the remark this answers is not in the document]
Comment (rdhananjaya) [reply tests/jballerina-integration-test/src/test/resources/grpc/src/grpcservices/03_bidirectional_streaming_service.bal]: @wggihan can you please assign appropriate person to review these stdlib changes.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java (hunk 1 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java]: What I meant was that we seem to be checking `isStructuredAccessOnVariableReference()` in both `isMutableVariable()` 
https://github.com/ballerina-platform/ballerina-lang/pull/19155/files#diff-25120e77c4c9017615d43d8e159cb3baR650
 and `isGlobalVarOrServiceVar()` https://github.com/ballerina-platform/ballerina-lang/pull/19155/files#diff-25120e77c4c9017615d43d8e159cb3baR684

Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java]: Got it, 
`isStructuredAccessOnVariableReference` is a check to decide weather to perform the logic in isMutableVariable and isGlobalVarOrServiceVar or to go one step in and perform the logic. I found it more readable to extract the checks into a function hence `isStructuredAccessOnVariableReference` was born.

**Pull request merged**